pytorch_lightning/loops/epoch/evaluation_epoch_loop.py

```diff
@@ -174,11 +181,20 @@ def on_save_checkpoint(self) -> Dict:
         state: Optional[MergedIteratorState] = getattr(self._data_fetcher.dataloader_iter, state_to_save, None)
         if state:
             state_dict["dataloader_state_dict"] = asdict(state)
+        state_dict["dataloader_state_dict"] = _collect_states_on_rank_zero_over_collection(
+            state_dict["dataloader_state_dict"], device=self.trainer.training_type_plugin.root_device
+        )
         return state_dict
 
     def on_load_checkpoint(self, state_dict: Dict) -> None:
         # cache the dataloader state dict until the dataloader objects are available
-        self._dataloader_state_dict = state_dict.get("dataloader_state_dict")
+        # dataset states are collected across all ranks
+        if _fault_tolerant_training():
+            dataloader_state_dict = state_dict.get("dataloader_state_dict", None)
+            if not dataloader_state_dict:
+                return
+            rank = torch.distributed.get_rank() if distributed_available() else 0
+            self._dataloader_state_dict = dataloader_state_dict[rank]
```

Comment:
isn't it possible to do this before we call the `on_load_checkpoint` hook? I believe this shouldn't be only loop specific

Reply:
Yes, I was thinking about this too. And I believed the cleanest was there. 

If you do it within the base loop, the user won't be its full state_dict and this logic is hidden from the user.

Reply:
I wouldn't do it in the base loop, but in the checkpoint connector on the full state dict, before we send it to the objects to restore. This line here

`self._dataloader_state_dict = dataloader_state_dict[rank]
`
tells me we can filter the dict already earlier before giving it to the objects. Isn't this better?

Reply:
I am afraid this would break the paradigm where state_dict should not being altered through connecting components such as the CheckpointConnector. I am personally not in favour of this and prefer to have the logic visible the reader.

This might also result in unwanted behaviour for users subclassing the loops and overriding those functions. How do you detect and automate the extraction of rank state dict in this case ?

@carmocca @ananthsub @justusschock Any thoughts ?

Reply:
I think it's fine to have the loop call the collect function, but the rest should be made strategy agnostic as much as possible. For example, instead of `torch.distributed.get_rank()` use call Lightning to determine the global_rank.